File: Backend/app/services/firestore_statistics_service.py
"""
Firestore Statistics Service for real-time user travel statistics.

This service connects to Firebase Firestore to retrieve real-time statistics
about user trips, plans, and expenses.
"""

import firebase_admin
from firebase_admin import credentials, firestore
from datetime import datetime
from typing import Dict, List, Optional
import logging

logger = logging.getLogger(__name__)

class FirestoreStatisticsService:
    """
    Service to retrieve and aggregate user statistics from Firestore.
    """
    def __init__(self):
        """
        Initialize the Firestore client.
        """
        self.db = firestore.client()
    
    def get_user_statistics(self, user_id: str) -> Dict:
        """
        Get aggregated statistics for a specific user from Firestore.
        
        Args:
            user_id (str): The Firebase UID of the user.
            
        Returns:
            Dict: A dictionary containing various statistics:
                - total_plans (int): Total number of plans created.
                - total_days (int): Total number of days across all plans (completed + ongoing).
                - locations_visited (int): Number of actual location check-ins.
                - total_trips (int): Number of completed trips.
                - total_expenses (float): Total actual expenses.
                - total_plans_2025 (int): Plans in 2025.
                - total_days_2025 (int): Days traveled in 2025.
                - locations_visited_2025 (int): Locations visited in 2025.
                - total_trips_2025 (int): Trips completed in 2025.
                - error (str): Error message if an exception occurs.
        """
        try:
            stats = {
                'total_plans': 0,
                'total_days': 0,
                'locations_visited': 0,
                'total_trips': 0,
                'total_expenses': 0.0,
                # Stats for 2025
                'total_plans_2025': 0,
                'total_days_2025': 0,
                'locations_visited_2025': 0,
                'total_trips_2025': 0
            }
            
            # Get user trips
            trips_ref = self.db.collection('users').document(user_id).collection('trips')
            trips = trips_ref.stream()
            
            now = datetime.now()
            year_2025_start = datetime(2025, 1, 1)
            year_2025_end = datetime(2025, 12, 31, 23, 59, 59)
            
            for trip_doc in trips:
                trip_data = trip_doc.to_dict()
                stats['total_plans'] += 1
                
                # Parse dates
                start_date = self._parse_date_helper(trip_data.get('startDate'))
                end_date = self._parse_date_helper(trip_data.get('endDate'))
                created_at = self._parse_date_helper(trip_data.get('createdAt')) or self._parse_date_helper(trip_data.get('created_at'))
                
                # Check if trip/plan is in 2025
                is_in_2025 = False
                if start_date and end_date:
                    # Check if trip overlaps with 2025
                    is_in_2025 = (end_date > year_2025_start and start_date < year_2025_end)
                elif created_at:
                    # Fallback: check creation date
                    is_in_2025 = created_at.year == 2025
                
                if is_in_2025:
                    stats['total_plans_2025'] += 1
                
                # Calculate trip duration and status
                if start_date and end_date:
                    duration = (end_date - start_date).days + 1
                    
                    # Check if trip is completed
                    if end_date < now:
                        stats['total_trips'] += 1
                        stats['total_days'] += duration  # Only count completed trips for total days
                        if is_in_2025:
                            stats['total_trips_2025'] += 1
                            stats['total_days_2025'] += duration
                    elif start_date <= now <= end_date:
                        # Ongoing trip - count days elapsed
                        days_elapsed = (now - start_date).days + 1
                        stats['total_days'] += days_elapsed
                        if is_in_2025:
                            stats['total_days_2025'] += days_elapsed
                
                # Count check-ins (locations actually visited)
                activities = trip_data.get('activities', [])
                for activity in activities:
                    # Check if activity is checked-in OR completed
                    is_checked_in = activity.get('checkIn', False) or activity.get('check_in', False)
                    is_completed = activity.get('status') == 'completed'
                    
                    # Filter by location-based activity types
                    activity_type = activity.get('activityType') or activity.get('activity_type', '')
                    is_location_based = self._is_location_based_activity(activity_type)
                    
                    if (is_checked_in or is_completed) and is_location_based:
                        stats['locations_visited'] += 1
                        if is_in_2025:
                            stats['locations_visited_2025'] += 1
                        # Add actual cost if available
                        if activity.get('actualCost'):
                            stats['total_expenses'] += float(activity.get('actualCost', 0))
            
            # Also check expenses collection
            try:
                expenses_ref = self.db.collection('users').document(user_id).collection('expenses')
                expenses = expenses_ref.stream()
                
                expenses_total = 0.0
                for expense_doc in expenses:
                    expense_data = expense_doc.to_dict()
                    actual_amount = expense_data.get('actual_amount', 0)
                    if actual_amount:
                        expenses_total += float(actual_amount)
                
                # Use expenses total if it's greater than activity costs
                if expenses_total > stats['total_expenses']:
                    stats['total_expenses'] = expenses_total
                    
            except Exception as e:
                logger.warning(f"Error getting expenses for user {user_id}: {e}")
            
            logger.info(f"Statistics for user {user_id}: {stats}")
            return stats
            
        except Exception as e:
            logger.error(f"Error getting statistics for user {user_id}: {e}")
            return {
                'total_plans': 0,
                'total_days': 0,
                'locations_visited': 0,
                'total_trips': 0,
                'total_expenses': 0.0,
                'total_plans_2025': 0,
                'total_days_2025': 0,
                'locations_visited_2025': 0,
                'total_trips_2025': 0,
                'error': str(e)
            }
    
    def _parse_date_helper(self, date_value) -> Optional[datetime]:
        """
        Helper method to parse different date formats (Firestore Timestamp, ISO string, etc.).

        Args:
            date_value: The date value to parse.

        Returns:
            Optional[datetime]: The parsed datetime object, or None if parsing fails.
        """
        if date_value is None:
            return None
        
        try:
            if hasattr(date_value, 'timestamp'):
                # Firestore Timestamp
                return datetime.fromtimestamp(date_value.timestamp())
            elif isinstance(date_value, str):
                # ISO string
                return datetime.fromisoformat(date_value.replace('Z', '+00:00')).replace(tzinfo=None)
            elif isinstance(date_value, datetime):
                # Already datetime
                return date_value.replace(tzinfo=None) if date_value.tzinfo else date_value
        except Exception as e:
            logger.warning(f"Error parsing date {date_value}: {e}")
        
        return None
    
    def _is_location_based_activity(self, activity_type: str) -> bool:
        """
        Check if an activity type represents a physical location that can be visited.

        Args:
            activity_type (str): The type of the activity.

        Returns:
            bool: True if the activity is location-based, False otherwise.
        """
        location_based_types = {
            'activity',
            'lodging', 
            'restaurant',
            'tour',
            'concert',
            'theater',
            'meeting',
            'parking'
        }
        return activity_type in location_based_types
    
    def get_trip_details(self, user_id: str) -> Dict:
        """
        Get detailed information about a user's trips (total, completed, ongoing, upcoming).
        
        Args:
            user_id (str): The Firebase UID of the user.

        Returns:
            Dict: A dictionary containing trip counts:
                - total (int)
                - completed (int)
                - ongoing (int)
                - upcoming (int)
                - error (str): Error message if applicable.
        """
        try:
            details = {
                'total': 0,
                'completed': 0,
                'ongoing': 0,
                'upcoming': 0
            }
            
            trips_ref = self.db.collection('users').document(user_id).collection('trips')
            trips = trips_ref.stream()
            
            now = datetime.now()
            
            for trip_doc in trips:
                trip_data = trip_doc.to_dict()
                details['total'] += 1
                
                start_date = trip_data.get('startDate')
                end_date = trip_data.get('endDate')
                
                if start_date and end_date:
                    try:
                        # Parse dates
                        if isinstance(start_date, str):
                            start_date = datetime.fromisoformat(start_date.replace('Z', '+00:00')).replace(tzinfo=None)
                        elif hasattr(start_date, 'timestamp'):
                            start_date = datetime.fromtimestamp(start_date.timestamp())
                            
                        if isinstance(end_date, str):
                            end_date = datetime.fromisoformat(end_date.replace('Z', '+00:00')).replace(tzinfo=None)
                        elif hasattr(end_date, 'timestamp'):
                            end_date = datetime.fromtimestamp(end_date.timestamp())
                        
                        # Categorize trip
                        if end_date < now:
                            details['completed'] += 1
                        elif start_date > now:
                            details['upcoming'] += 1
                        else:
                            details['ongoing'] += 1
                            
                    except Exception as e:
                        logger.warning(f"Error parsing dates for trip {trip_doc.id}: {e}")
            
            return details
            
        except Exception as e:
            logger.error(f"Error getting trip details for user {user_id}: {e}")
            return {'total': 0, 'completed': 0, 'ongoing': 0, 'upcoming': 0, 'error': str(e)}
    
    def get_monthly_expenses(self, user_id: str, limit: int = 12) -> List[Dict]:
        """
        Get monthly expense totals for a user.
        
        Args:
            user_id (str): The Firebase UID of the user.
            limit (int): Maximum number of months to return. Defaults to 12.

        Returns:
            List[Dict]: A list of dictionaries, each containing:
                - month (str): The month key (YYYY-MM).
                - amount (float): Total expense amount for that month.
        """
        try:
            monthly_expenses = {}
            
            # Get from expenses collection
            expenses_ref = self.db.collection('users').document(user_id).collection('expenses')
            expenses = expenses_ref.order_by('created_at', direction=firestore.Query.DESCENDING).limit(50).stream()
            
            for expense_doc in expenses:
                expense_data = expense_doc.to_dict()
                created_at = expense_data.get('created_at')
                actual_amount = expense_data.get('actual_amount', 0)
                
                if created_at and actual_amount:
                    if hasattr(created_at, 'timestamp'):
                        date = datetime.fromtimestamp(created_at.timestamp())
                    else:
                        continue
                        
                    month_key = f"{date.year}-{date.month:02d}"
                    monthly_expenses[month_key] = monthly_expenses.get(month_key, 0) + float(actual_amount)
            
            # Convert to list and sort
            result = [
                {'month': month, 'amount': amount}
                for month, amount in monthly_expenses.items()
            ]
            result.sort(key=lambda x: x['month'])
            
            return result[-limit:] if len(result) > limit else result
            
        except Exception as e:
            logger.error(f"Error getting monthly expenses for user {user_id}: {e}")
            return []

File: lib/Login/services/firestore_statistics_service.dart
import 'dart:async';
import 'package:cloud_firestore/cloud_firestore.dart';
import '../../Login/services/auth_service.dart';
import 'package:flutter/foundation.dart';

/// Model for real-time user travel statistics from Firestore
class UserTravelStats {
  final int totalActivities;
  final int completedTrips;
  final int checkedInLocations;
  final double totalExpenses;
  final int totalPlans;
  final int ongoingTrips;
  final int upcomingTrips;
  final Map<String, double> monthlyExpenses;
  final double averageExpensePerTrip;
  final int totalDays;
  final double totalDistance;
  
  // Stats for 2025 only
  final int totalActivities2025;
  final int completedTrips2025;
  final int checkedInLocations2025;
  final int totalPlans2025;
  final int totalDays2025;

  UserTravelStats({
    required this.totalActivities,
    required this.completedTrips,
    required this.checkedInLocations,
    required this.totalExpenses,
    required this.totalPlans,
    required this.ongoingTrips,
    required this.upcomingTrips,
    required this.monthlyExpenses,
    required this.averageExpensePerTrip,
    required this.totalDays,
    required this.totalDistance,
    required this.totalActivities2025,
    required this.completedTrips2025,
    required this.checkedInLocations2025,
    required this.totalPlans2025,
    required this.totalDays2025,
  });

  factory UserTravelStats.empty() {
    return UserTravelStats(
      totalActivities: 0,
      completedTrips: 0,
      checkedInLocations: 0,
      totalExpenses: 0.0,
      totalPlans: 0,
      ongoingTrips: 0,
      upcomingTrips: 0,
      monthlyExpenses: {},
      averageExpensePerTrip: 0.0,
      totalDays: 0,
      totalDistance: 0.0,
      totalActivities2025: 0,
      completedTrips2025: 0,
      checkedInLocations2025: 0,
      totalPlans2025: 0,
      totalDays2025: 0,
    );
  }

  UserTravelStats copyWith({
    int? totalActivities,
    int? completedTrips,
    int? checkedInLocations,
    double? totalExpenses,
    int? totalPlans,
    int? ongoingTrips,
    int? upcomingTrips,
    Map<String, double>? monthlyExpenses,
    double? averageExpensePerTrip,
    int? totalDays,
    double? totalDistance,
    int? totalActivities2025,
    int? completedTrips2025,
    int? checkedInLocations2025,
    int? totalPlans2025,
    int? totalDays2025,
  }) {
    return UserTravelStats(
      totalActivities: totalActivities ?? this.totalActivities,
      completedTrips: completedTrips ?? this.completedTrips,
      checkedInLocations: checkedInLocations ?? this.checkedInLocations,
      totalExpenses: totalExpenses ?? this.totalExpenses,
      totalPlans: totalPlans ?? this.totalPlans,
      ongoingTrips: ongoingTrips ?? this.ongoingTrips,
      upcomingTrips: upcomingTrips ?? this.upcomingTrips,
      monthlyExpenses: monthlyExpenses ?? this.monthlyExpenses,
      averageExpensePerTrip: averageExpensePerTrip ?? this.averageExpensePerTrip,
      totalDays: totalDays ?? this.totalDays,
      totalDistance: totalDistance ?? this.totalDistance,
      totalActivities2025: totalActivities2025 ?? this.totalActivities2025,
      completedTrips2025: completedTrips2025 ?? this.completedTrips2025,
      checkedInLocations2025: checkedInLocations2025 ?? this.checkedInLocations2025,
      totalPlans2025: totalPlans2025 ?? this.totalPlans2025,
      totalDays2025: totalDays2025 ?? this.totalDays2025,
    );
  }
}

/// Service for getting real-time statistics from Firestore
class FirestoreStatisticsService {
  static final FirestoreStatisticsService _instance = FirestoreStatisticsService._internal();
  factory FirestoreStatisticsService() => _instance;
  FirestoreStatisticsService._internal();

  final FirebaseFirestore _firestore = FirebaseFirestore.instance;
  final AuthService _authService = AuthService();
  
  StreamController<UserTravelStats>? _statsController;
  StreamSubscription<QuerySnapshot>? _tripsSubscription;
  StreamSubscription<QuerySnapshot>? _expensesSubscription;

  /// Get real-time statistics stream
  Stream<UserTravelStats> getUserStatisticsStream() {
    final currentUser = _authService.currentUser;
    if (currentUser == null) {
      return Stream.value(UserTravelStats.empty());
    }

    _statsController?.close();
    _statsController = StreamController<UserTravelStats>.broadcast();

    _listenToTripsAndExpenses(currentUser.uid);
    
    return _statsController!.stream;
  }

  /// Listen to trips and expenses changes
  void _listenToTripsAndExpenses(String userId) {
    UserTravelStats currentStats = UserTravelStats.empty();
    bool tripsLoaded = false;
    bool expensesLoaded = false;

    // Emit empty stats immediately to show UI quickly
    _statsController?.add(currentStats);

    // Listen to trips changes - filter for private trips in processing
    _tripsSubscription = _firestore
        .collection('users')
        .doc(userId)
        .collection('trips')
        .snapshots()
        .listen((tripsSnapshot) {
      final tripsStats = _calculateTripsStatistics(tripsSnapshot);
      currentStats = currentStats.copyWith(
        totalActivities: tripsStats['totalActivities'],
        completedTrips: tripsStats['completedTrips'],
        ongoingTrips: tripsStats['ongoingTrips'],
        upcomingTrips: tripsStats['upcomingTrips'],
        totalPlans: tripsStats['totalPlans'],
        checkedInLocations: tripsStats['checkedInLocations'],
        totalDays: tripsStats['totalDays'],
        totalDistance: tripsStats['totalDistance'],
        totalActivities2025: tripsStats['totalActivities2025'],
        completedTrips2025: tripsStats['completedTrips2025'],
        totalPlans2025: tripsStats['totalPlans2025'],
        checkedInLocations2025: tripsStats['checkedInLocations2025'],
        totalDays2025: tripsStats['totalDays2025'],
      );
      
      tripsLoaded = true;
      // Emit stats immediately when trips are loaded, don't wait for expenses
      _statsController?.add(currentStats);
    }, onError: (error) {
      debugPrint('Error loading trips: $error');
      tripsLoaded = true;
      _statsController?.add(currentStats);
    });

    // Listen to expenses changes
    _expensesSubscription = _firestore
        .collection('users')
        .doc(userId)
        .collection('expenses')
        .snapshots()
        .listen((expensesSnapshot) {
      final expensesStats = _calculateExpensesStatistics(expensesSnapshot);
      
      // Calculate average expense per trip if we have completed trips
      final completedTrips = currentStats.completedTrips > 0 ? currentStats.completedTrips : 1;
      final averageExpensePerTrip = expensesStats['totalExpenses'] / completedTrips;
      
      currentStats = currentStats.copyWith(
        totalExpenses: expensesStats['totalExpenses'],
        monthlyExpenses: expensesStats['monthlyExpenses'],
        averageExpensePerTrip: averageExpensePerTrip,
      );
      
      expensesLoaded = true;
      // Always emit updated stats when expenses are loaded
      _statsController?.add(currentStats);
    }, onError: (error) {
      debugPrint('Error loading expenses: $error');
      expensesLoaded = true;
      // Continue without expenses data
      _statsController?.add(currentStats);
    });
  }

  /// Calculate statistics from trips
  Map<String, dynamic> _calculateTripsStatistics(QuerySnapshot snapshot) {
    int totalActivities = 0;
    int completedTrips = 0;
    int ongoingTrips = 0;
    int upcomingTrips = 0;
    int totalPlans = 0; // Will count only private plans
    int checkedInLocations = 0;
    int totalDays = 0;
    double totalDistance = 0.0;

    // Stats for 2025 only
    int totalActivities2025 = 0;
    int completedTrips2025 = 0;
    int totalPlans2025 = 0;
    int checkedInLocations2025 = 0;
    int totalDays2025 = 0;

    final now = DateTime.now();
    final year2025Start = DateTime(2025, 1, 1);
    final year2025End = DateTime(2025, 12, 31, 23, 59, 59);

    for (final doc in snapshot.docs) {
      final data = doc.data() as Map<String, dynamic>;
      
      // Only count private trips (trips with no collaborators)
      final collaborators = data['collaborators'] as List<dynamic>? ?? [];
      if (collaborators.isNotEmpty) {
        continue; // Skip shared/collaborative trips
      }
      
      // Count this as a private plan
      totalPlans++;
      
      // Parse dates
      final startDate = _parseDate(data['startDate']);
      final endDate = _parseDate(data['endDate']);
      final createdAt = _parseDate(data['createdAt']) ?? _parseDate(data['created_at']);

      // Check if trip/plan is created in 2025 or overlaps with 2025
      bool isIn2025 = false;
      if (startDate != null && endDate != null) {
        // Check if trip overlaps with 2025 (starts before 2026 and ends after 2024)
        isIn2025 = endDate.isAfter(year2025Start.subtract(Duration(days: 1))) && 
                   startDate.isBefore(year2025End.add(Duration(days: 1)));
      } else if (createdAt != null) {
        // Fallback: check creation date
        isIn2025 = createdAt.year == 2025;
      }

      // Count activities
      final activities = data['activities'] as List<dynamic>? ?? [];
      totalActivities += activities.length;
      if (isIn2025) {
        totalActivities2025 += activities.length;
        totalPlans2025++;
      }

      // Count check-ins (locations visited where user actually went) and analyze trip completion
      List<Map<String, dynamic>> locationBasedActivities = [];
      int checkedInActivitiesInTrip = 0;
      
      for (final activity in activities) {
        if (activity is Map<String, dynamic>) {
          final activityType = activity['activityType'] ?? activity['activity_type'] ?? '';
          final isLocationBased = _isLocationBasedActivity(activityType);
          
          if (isLocationBased) {
            locationBasedActivities.add(activity);
            
            // Check if activity is checked-in OR completed
            final isCheckedIn = activity['checkIn'] == true || activity['check_in'] == true;
            final isCompleted = activity['status'] == 'completed';
            
            if (isCheckedIn || isCompleted) {
              checkedInLocations++;
              checkedInActivitiesInTrip++;
              if (isIn2025) {
                checkedInLocations2025++;
              }
            }
          }
        }
      }

      // Check if trip is completed based on check-ins
      // A trip is completed only when ALL location-based activities are checked-in
      bool isTripCompleted = locationBasedActivities.isNotEmpty && 
                           checkedInActivitiesInTrip == locationBasedActivities.length;

      // Categorize trips by completion status (not by date)
      if (startDate != null && endDate != null) {
        // Calculate trip duration in days  
        final duration = endDate.difference(startDate).inDays + 1;
        
        if (isTripCompleted) {
          // Trip is completed (all locations checked in)
          completedTrips++;
          if (isIn2025) {
            completedTrips2025++;
          }
        } else if (startDate.isAfter(now)) {
          upcomingTrips++;
        } else if (endDate.isBefore(now)) {
          // Trip date passed but not all activities checked in - still counts as incomplete
          // Don't count as completed trip
        } else {
          // Trip is ongoing
          ongoingTrips++;
        }

        // Estimate distance based on trip duration (example: 100km per day)
        totalDistance += duration * 100.0;
      }
    }

    return {
      'totalActivities': totalActivities,
      'completedTrips': completedTrips,
      'ongoingTrips': ongoingTrips,
      'upcomingTrips': upcomingTrips,
      'totalPlans': totalPlans,
      'checkedInLocations': checkedInLocations,
      'totalDays': totalDays,
      'totalDistance': totalDistance,
      'totalActivities2025': totalActivities2025,
      'completedTrips2025': completedTrips2025,
      'totalPlans2025': totalPlans2025,
      'checkedInLocations2025': checkedInLocations2025,
      'totalDays2025': totalDays2025,
    };
  }

  /// Check if activity type represents a location that can be visited
  bool _isLocationBasedActivity(String activityType) {
    const locationBasedTypes = {
      'activity',
      'lodging', 
      'restaurant',
      'tour',
      'concert',
      'theater',
      'meeting',
      'parking'
    };
    return locationBasedTypes.contains(activityType);
  }

  /// Calculate statistics from expenses
  Map<String, dynamic> _calculateExpensesStatistics(QuerySnapshot snapshot) {
    double totalExpenses = 0.0;
    Map<String, double> monthlyExpenses = {};

    for (final doc in snapshot.docs) {
      final data = doc.data() as Map<String, dynamic>;
      
      final actualAmount = (data['actual_amount'] ?? data['actualAmount'] ?? 0).toDouble();
      totalExpenses += actualAmount;

      // Group by month
      final createdAt = _parseDate(data['created_at'] ?? data['createdAt']);
      if (createdAt != null) {
        final monthKey = '${createdAt.year}-${createdAt.month.toString().padLeft(2, '0')}';
        monthlyExpenses[monthKey] = (monthlyExpenses[monthKey] ?? 0) + actualAmount;
      }
    }

    // Calculate average expense per trip (will be updated when trips data is available)
    double averageExpensePerTrip = 0.0;

    return {
      'totalExpenses': totalExpenses,
      'monthlyExpenses': monthlyExpenses,
      'averageExpensePerTrip': averageExpensePerTrip,
    };
  }

  /// Parse date from various formats
  DateTime? _parseDate(dynamic date) {
    if (date == null) return null;
    
    try {
      if (date is Timestamp) {
        return date.toDate();
      } else if (date is String) {
        return DateTime.parse(date);
      } else if (date is DateTime) {
        return date;
      }
    } catch (e) {
      debugPrint('Error parsing date: $e');
    }
    
    return null;
  }

  /// Get statistics once (not real-time)
  Future<UserTravelStats> getUserStatistics() async {
    final currentUser = _authService.currentUser;
    if (currentUser == null) {
      return UserTravelStats.empty();
    }

    try {
      // Get trips - filter for private trips in processing
      final tripsSnapshot = await _firestore
          .collection('users')
          .doc(currentUser.uid)
          .collection('trips')
          .get();

      // Get expenses
      final expensesSnapshot = await _firestore
          .collection('users')
          .doc(currentUser.uid)
          .collection('expenses')
          .get();

      final tripsStats = _calculateTripsStatistics(tripsSnapshot);
      final expensesStats = _calculateExpensesStatistics(expensesSnapshot);

      // Calculate average expense per trip
      final averageExpensePerTrip = tripsStats['completedTrips'] > 0
          ? expensesStats['totalExpenses'] / tripsStats['completedTrips']
          : 0.0;

      return UserTravelStats(
        totalActivities: tripsStats['totalActivities'],
        completedTrips: tripsStats['completedTrips'],
        ongoingTrips: tripsStats['ongoingTrips'],
        upcomingTrips: tripsStats['upcomingTrips'],
        totalPlans: tripsStats['totalPlans'],
        checkedInLocations: tripsStats['checkedInLocations'],
        totalDays: tripsStats['totalDays'],
        totalDistance: tripsStats['totalDistance'],
        totalExpenses: expensesStats['totalExpenses'],
        monthlyExpenses: expensesStats['monthlyExpenses'],
        averageExpensePerTrip: averageExpensePerTrip,
        totalActivities2025: tripsStats['totalActivities2025'],
        completedTrips2025: tripsStats['completedTrips2025'],
        totalPlans2025: tripsStats['totalPlans2025'],
        checkedInLocations2025: tripsStats['checkedInLocations2025'],
        totalDays2025: tripsStats['totalDays2025'],
      );
    } catch (e) {
      debugPrint('Error getting user statistics: $e');
      return UserTravelStats.empty();
    }
  }

  /// Dispose resources
  void dispose() {
    _tripsSubscription?.cancel();
    _expensesSubscription?.cancel();
    _statsController?.close();
  }
}